apps/system/js/audio_channel_manager.js

```diff
@@ -0,0 +1,657 @@
+   * like Music app would like to play music when FM app is already playing
+   * or a audio channel playing in background.
+   *
+   * @example
+   * Service.request('registerAudioChannel', app);
+   *
+   * @class AudioChannelManager
+   */
+  var AudioChannelManager = function() {};
+
+  AudioChannelManager.SERVICES = [
+    'registerAudioChannel',
+    'unregisterAudioChannel'
+  ];
+
+  BaseModule.create(AudioChannelManager, {
+    name: 'AudioChannelManager',
+    DEBUG: false,
+    _isVibrateEnabled: true,
+    /**
+     * The array element is an `AppAudioChannel` object.
+     *
+     * @example
+     * [{ appId: 'String', channel: BrowserElementAudioChannel }]
+     */
+    _activeAudioChannels: null,
+    /**
+     * The array element is an `InactiveAppAudioChannel` object.
+     *
+     * @example
+     * [{
+     *   appId: 'String',
+     *   channel: BrowserElementAudioChannel,
+     *   resumeConditions: {
+     *     // Wait for the app audio channel ends.
+     *     interruptionAppAduioChannel: {
+     *       appId: 'String',
+     *       channel: BrowserElementAudioChannel 
+     *     },
+     *     resumeWhenForeground: Boolean,
+     *     withVibrate: Boolean,
+     *     vibrationIntervalId: 'String'
+     *   }
+     * }]
+     */
+    _inactiveAudioChannels: null,
+    /**
+     * The array element is a `FadedAppAudioChannel` object.
+     *
+     * @example
+     * [{
+     *   appId: 'String',
+     *   audioChannel: BrowserElementAudioChannel,
+     *   // Wait for the app audio channel ends.
+     *   interruptionAudioChannel: appAudioChannel
+     * }]
+     */
+    _fadedAudioChannels: null,
+
+    /**
+     * Initial the module.
+     */
+    _start: function() {
+      this._activeAudioChannels = [];
+      this._inactiveAudioChannels = [];
+      this._fadedAudioChannels = [];
+      window.addEventListener('appopened', this);
+      window.addEventListener('appclosed', this);
+      var settings = navigator.mozSettings;
+      var vibration = settings.createLock().get('vibration.enabled');
+      vibration.onsuccess = () => {
+        this._isVibrateEnabled = vibration.result['vibration.enabled'];
+      };
+      vibration.onerror = () => {
+        this.debug('Error of getting vibration.enabled: ' + vibration.error);
+      };
+      settings.addObserver('vibration.enabled', (evt) => {
+        this._isVibrateEnabled = evt.settingValue;
+      });
+      this.debug('Start the audio channel manager module.');
+    },
+
+    /**
+     * Add an app into manager to manage its audio channels.
+     *
+     * @param {AppWindow} app An app you want to manage.
+     */
+    registerAudioChannel: function(app) {
+      var allowedAudioChannels = app.browser.element.allowedAudioChannels;
+      // Feature detection for using the new audio channel management module.
+      allowedAudioChannels && allowedAudioChannels.forEach((channel) => {
```

Comment:
After thinking, I'd like to see this part being a subcomponent of each appWindow in order to make manager more simple.

Below is pseudo code
```js
var AudioChannelController = function(app) {
  this.app = app;
  this.app.browser.element.allowedAudioChannels.forEach(function(channel) {
    channel.addEventListener('activestatechanged', this);
  }, this);
};
AudioChannelController.prototype = Object.create(BaseUI.prototype);
AudioChannelController.prototype = {
  handleEvent: function(evt) {
    this.publish('audiochannelactivated', { channel: evt.target, app: this.app });
  }
};
```
```js
AudioChannelManager.EVENTS = [
  'audiochannelstatechanged'
];
BaseModule.create(AudioChannelManager, {
  name: 'AudioChannelManager',
  _handle_audiochannelstatechanged: function(evt) {
    if (evt.channel.isActive()) {
      this._handleAudioChannel('competition', appId, channel);
    } else {
      this._resetAudioChannel(appId, channelName);
      this._resetFadedAudioChannelBy(appId, channelName);
      this._resumeInterruptedAudioChannelBy({
         interruptionAppAduioChannel: {
            appId: appId,
            channel: channel
          }
       });
    }
  }
});
```